### PR TITLE
SI-7039 unapplySeq result type independent of subpattern count

### PR DESCRIPTION
**Pull request by adriaanm**
Fixes a bug in the implementation of the `unapplySeq` part of the spec below.

An `unapply` method with result type `R` in an object `x` matches the 
pattern `x(p_1, ..., p_n)` if it takes exactly one argument and, either:
- `n = 0` and `R =:= Boolean`, or
- `n = 1` and `R <:< Option[T]`, for some type `T`.
   The argument pattern `p1` is typed in turn with expected type `T`.
- Or, `n > 1` and `R <:< Option[Product_n[T_1, ..., T_n]]`, for some
  types `T_1, ..., T_n`. The argument patterns `p_1, ..., p_n` are
  typed with expected types `T_1, ..., T_n`.

An `unapplySeq` method in an object `x` matches the pattern `x(p_1, ..., p_n)`
if it takes exactly one argument and its result type is of the form `Option[S]`,
where either:
- `S` is a subtype of `Seq[U]` for some element type `U`, (set `m = 0`)
- or `S` is a `ProductX[T_1, ..., T_m]` and `T_m <: Seq[U]` (`m <= n`).

The argument patterns `p_1, ..., p_n` are typed with expected types
`T_1, ..., T_m, U, ..., U`. Here, `U` is repeated `n-m` times.

review by @retronym, /cc @jrudolph (thanks for the original patch in #2004!)

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -83,31 +94,32 @@ trait Infer extends Checkable {
...
+    def productArgs = getProductArgs(optionArgs.head)
+
     val formals =
-      if (nbSubPats == 0 && booleanExtractor && !isUnapplySeq)  Nil
-      else resTp.baseType(OptionClass).typeArgs match {
-        case optionTArg :: Nil =>
-          def productArgs = getProductArgs(optionTArg)
+      if (isUnapplySeq) {
+        if (optionArgs.nonEmpty)
+          productArgs match {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: `case initTps :+ seqTp => initTps :+ seqToRepeatedChecked(seqTp)` is a fraction prettier.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -83,31 +94,32 @@ trait Infer extends Checkable {
...
+            case Nil => List(seqToRepeatedChecked(optionArgs.head))
+            case tps => tps.init :+ seqToRepeatedChecked(tps.last)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: `s/unappSym.owner + unappSym.owner.locationString/unappSym.fullLocationString/`. This error is dup-ed below, it ought be extracted.

File: test/files/run/t7039.scala [hunk 3]
```diff
@@ -0,0 +1,10 @@
+object UnapplySeqTest {
+  def unapplySeq(any: Any): Option[(Int, Seq[Int])] = Some((5, List(1)))
+}
+
+object Test extends App {
+  null match {
+    case UnapplySeqTest(5) => println("uh-oh")
+    case UnapplySeqTest(5, 1) => println("Matched!") // compiles
+  }
```
Comment (retronym) [inline test/files/run/t7039.scala]: Maybe throw in a `case UnapplySeqTest(5, xs @ _*) => xs: Seq[Int] }` for good measure.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: it's not prettier -- it's beauteous!
didn't even realize you could do that :eyes:

**Comment by adriaanm** [comment]
the kitten has been fixed -- it'll rebuild when it notices later commits

**Comment by adriaanm** [comment]
also, working on fixing your feedback, so it'll have to restart all over anyway in a minute

**Comment by adriaanm** [comment]
rebased to address review comments

**Comment by adriaanm** [comment]
ah yes, I don't think the logic knows much about list lengths :-(

**Comment by adriaanm** [comment]
the last case (in run/t7039 as it stands) must be considered reachable, though, since the result type of the unapplySeq doesn't specify the second element must be a `1`, nor that the seq must be of length 1

**Comment by adriaanm** [comment]
ping @retronym for final LGTM

**Comment by retronym** [comment]
you got it, lgtm

**Comment by adriaanm** [comment]
thanks -- btw, the kitten now recognizes LGTM's if they start the commit and will add the "reviewed" label accordingly

**Pull request merged**